### DISCUSSION TITLE
Filtering core midi devices

### DISCUSSION THREAD

**Discussion by SNewtown** [orchetect/swift-midi#217] [Q&A]
Hi Steffan,

I now have my observable list of devices working on iOS using the latest version, which is great, but my intent being to list only connected devices, I also get “Bluetooth” and “Network” since those are classified as devices, which I filtered out by name.  But on devices running on different languages, I don’t think this approach works.  I did a bit of research and I might be able to find a solution if I query the device using core midi (as there seems to be additional properties that could potentially be used to identify these beyond the name), I was wondering if you have a more creative way to filter these out of my list of devices, that wouldn’t be impacted by regional context?

Cheers,
Stephane

**Comment by orchetect** [comment]
Any reason to filter them out? Or preclude a user from using them?

One way may be to filter by [driver owner](https://orchetect.github.io/MIDIKit/documentation/midikit/midiinputendpoint/driverowner). It ought to be stable across locales. You can get that property on any device, entity, or endpoint.

Bluetooth on macOS always has the owner ID `com.apple.AppleMIDIBluetoothDriver`, and Network always has the owner ID `com.apple.AppleMIDIRTPDriver`. for example. Not sure if it's the same on iOS but it's easy to check using the MIDISystemInfo example project on macOS or iOS.

**Reply by SNewtown** [reply]
Thanks Steffan!

I will definitely try the owner filter.  Basically my app is mostly about translating CC/PC inbound to SysEx outbound.  Hence I listen to all connected devices (for any inbound CC/PC) and once processed, send the resulting SysEx to all connected devices.  The reason for the listing of devices is just an added benefit for the user to see what's connected and since Bluetooth and Network are a given (and not "actual" devices) it just makes it a cleaner list on the UI.  As for managing the connections, the user is provided the Bluetooth connectivity screen so this is just a listing of what's actively connected (it doesn't preclude the use of anything).  Hope this additional context helps to position my inquiry and I'm sure filtering the owner will work just fine, but will report back once I give it a whirl

**Reply by SNewtown** [reply]
Turns out that Bluetooth is not renamed (at least not in the Spanish Locale) but Network is, which is a good thing because the driverOwner of the Bluetooth is the same as most connected devices... so my filter to show only connected devices and take out Bluetooth and Network from my list is this:

observableDevices.devices.filter { $0.isOffline == false  && $0.name != "Bluetooth" && $0.driverOwner != "com.apple.AppleMIDIRTPDriver" }

And that seems to work (at least in both English and Spanish locales).  The risk of this not working in other locales is relatively inconsequential, those users would simply see an additional device with whatever name is given to Network (if it has a different driverOwner) or Bluetooth (if that is renamed in that locale).

I think I'm fine with that, removing these two is simply to declutter the view of "physical" connected devices :)

Thanks again Steffan!
